### PR TITLE
Exchanges integration storers

### PR DESCRIPTION
**Pull request by andreibancioiu**
1) After restarting a node, `GetFromEpoch` could not access persisters older than `NumActivePersisters`.

 - Fix: when `CleanOldEpochsData` is unset, then open all persisters (and close old ones, as before)
 - Fix: call `setIsClosed()` when closing the persisters

2) In the PR [2226](https://github.com/ElrondNetwork/elrond-go/pull/2226), I've mistakenly disabled `epoch by tx hash` indexing (and thus `get tx by hash` API wouldn't work for older epochs). Re-made now the link to epoch, by making `MiniblockHashByTxHashStorage` Static. **Put transactions is fast (1 put per transaction), get 1 transaction is slower (3 reads).**

3) Index genesis block (for hyperblock nonce 1).

4) Fixed path for EpochByHashStorage. `FilePath = "EpochByHash"`, it was empty due to PR [2226](https://github.com/ElrondNetwork/elrond-go/pull/2226).

5) Fixed edge-case for saving hyperblock coordinates of cross-epoch (committed in 42, notarized in 43) transactions. Using `putInEpoch` primitive.

6) Rename "fullHistory" to "db lookup extensions".

**How to test**

1 and 2 - wait a few epochs, stop some observers, start them again, query older transactions and older (hyper)blocks.

3 - Query hyperblock 1 (by nonce, by hash). Query shard blocks 0 (by nonce, by hash). Compare with Explorer.

4 - Assert that DB folders structure in `Static` is nice.

5 - A bit harder to test manually - we have unit tests though.

6 - Inspect `config.toml` and `node/db`'s structure.

**Note that, for tests, I've set**:

```
RoundsPerEpoch = 20
FullHistory.Enabled = true
```

### CODE REVIEW

**Review by miiu96 — DISMISSED** [review]

File: cmd/node/config/config.toml [hunk 1]
```diff
@@ -480,8 +480,8 @@
    Type = "json"
 
 [EpochStartConfig]
-    MinRoundsBetweenEpochs = 200
-    RoundsPerEpoch         = 200
+    MinRoundsBetweenEpochs = 20 # UNDO AFTER TESTS
```
Comment (iulianpascalau) [inline cmd/node/config/config.toml:483]: Might add a TODO as to signal the line

File: cmd/node/config/config.toml [hunk 2]
```diff
@@ -645,7 +645,7 @@
     PollingIntervalInMinutes = 65
 
 [FullHistory]
-    Enabled = false
+    Enabled = true # DISABLE AFTER TESTS
```
Comment (iulianpascalau) [inline cmd/node/config/config.toml:648]: TODO?

**Review by AdoAdoAdo — APPROVED** [review]

File: storage/pruning/pruningStorer.go [hunk 3]
```diff
@@ -201,8 +201,12 @@ func initPersistersInEpoch(
 	if oldestEpochActive < 0 {
 		oldestEpochActive = 0
 	}
+	if !args.CleanOldEpochsData {
+		oldestEpochKeep = 0
```
Comment (iulianpascalau) [inline storage/pruning/pruningStorer.go:205]: This solution is not quite extensible. If we have thousands of epochs, the node will soon run out of memory. Plus, as seen today opening a storage is a quite costly operation...on the order of magnitude of seconds.

File: storage/pruning/pruningStorer.go [hunk 4]
```diff
@@ -201,8 +201,12 @@ func initPersistersInEpoch(
...
+	}
 
 	for epoch := int64(args.StartingEpoch); epoch >= oldestEpochKeep; epoch-- {
+		log.Debug("initPersistersInEpoch(): createPersisterDataForEpoch", "identifier", args.Identifier, "epoch", epoch, "shardID", shardIDStr)
 		p, err := createPersisterDataForEpoch(args, uint32(epoch), shardIDStr)
```
Comment (AdoAdoAdo) [inline storage/pruning/pruningStorer.go:210]: talking with Iulian, seems that opening and closing 1000 DBs can take up to 20 minutes.
Maybe we can change this function createPersisterDataForEpoch, to only create the reference but not open/close. 
We should only open the active persisters at the end.

**Review by AdoAdoAdo — CHANGES_REQUESTED** [review]

File: cmd/node/config/config.toml (hunk 1 above)
Comment (andreibancioiu) [reply cmd/node/config/config.toml]: Added.

File: cmd/node/config/config.toml (hunk 2 above)
Comment (andreibancioiu) [reply cmd/node/config/config.toml]: Added.

File: storage/pruning/pruningStorer.go (hunk 3 above)
Comment (andreibancioiu) [reply storage/pruning/pruningStorer.go]: Sketched a solution.

File: storage/pruning/pruningStorer.go (hunk 4 above)
Comment (andreibancioiu) [reply storage/pruning/pruningStorer.go:215]: Sketched a solution.

File: storage/pruning/pruningStorer.go [hunk 5]
```diff
@@ -340,6 +348,7 @@ func (ps *PruningStorer) GetFromEpoch(key []byte, epoch uint32) ([]byte, error)
 	}
 
 	persister, err := ps.persisterFactory.Create(pd.path)
+	// Question for review: save reference on persister data?
```
Comment (iulianpascalau) [inline storage/pruning/pruningStorer.go:351]: IMHO, I do not think we will need the old closed reference.

File: storage/pruning/pruningStorer.go [hunk 6]
```diff
@@ -310,6 +316,8 @@ func (ps *PruningStorer) Close() error {
 			log.Error("cannot close persister", err)
 			closedSuccessfully = false
 		}
+		// Question for review - all good here?
+		persister.setIsClosed(true)
```
Comment (iulianpascalau) [inline storage/pruning/pruningStorer.go:320]: :+1:

File: storage/pruning/pruningStorer.go [hunk 7]
```diff
@@ -350,6 +359,8 @@ func (ps *PruningStorer) GetFromEpoch(key []byte, epoch uint32) ([]byte, error)
 		if err != nil {
 			log.Debug("persister.Close()", "error", err.Error())
 		}
+		// Question for review: all good here?
+		pd.setIsClosed(true)
```
Comment (iulianpascalau) [inline storage/pruning/pruningStorer.go:363]: :+1:

File: storage/pruning/pruningStorer.go [hunk 8]
```diff
@@ -400,6 +412,8 @@ func (ps *PruningStorer) GetBulkFromEpoch(keys [][]byte, epoch uint32) (map[stri
 			if err != nil {
 				log.Debug("persister.Close()", "error", err.Error())
 			}
+			// Question for review - all good here?
+			pd.setIsClosed(true)
```
Comment (iulianpascalau) [inline storage/pruning/pruningStorer.go:416]: :+1:

File: storage/pruning/pruningStorer.go [hunk 9]
```diff
@@ -517,6 +532,8 @@ func (ps *PruningStorer) HasInEpoch(key []byte, epoch uint32) error {
...
```
Comment (iulianpascalau) [inline storage/pruning/pruningStorer.go:536]: :+1:

File: storage/pruning/pruningStorer.go [hunk 10]
```diff
@@ -882,14 +899,33 @@ func (ps *PruningStorer) IsInterfaceNil() bool {
 	return ps == nil
 }
 
-func createPersisterDataForEpoch(args *StorerArgs, epoch uint32, shardIdStr string) (*persisterData, error) {
-	// TODO: if booting from storage in an epoch > 0, shardId needs to be taken from somewhere else
-	// e.g. determined from directories in persister path or taken from boot storer
+func createShallowPersisterDataForEpoch(args *StorerArgs, epoch uint32, shardIdStr string) *persisterData {
+	filePath := createPersisterPathForEpoch(args, epoch, shardIdStr)
+
+	p := &persisterData{
```
Comment (iulianpascalau) [inline storage/pruning/pruningStorer.go:905]: return directly?

**Review by bogdan-rosianu — APPROVED** [review]

File: storage/pruning/pruningStorer.go (hunk 10 above)
Comment (andreibancioiu) [reply storage/pruning/pruningStorer.go]: Fixed.

File: storage/pruning/pruningStorer.go [hunk 11]
```diff
@@ -245,11 +245,16 @@ func (ps *PruningStorer) Put(key, data []byte) error {
 			log.Debug("active persister not found",
 				"epoch", ps.epochForPutOperation,
 				"used", persisterToUse.epoch)
+
+			// Question for review: isn't this a critical error?
```
Comment (bogdan-rosianu) [inline storage/pruning/pruningStorer.go:249]: I think it is critical because otherwise you'd save data in a persister which is not for the desired epoch

**Review by bogdan-rosianu — APPROVED** [review]

File: storage/pruning/pruningStorer.go (hunk 11 above)
Comment (iulianpascalau) [reply storage/pruning/pruningStorer.go:249]: If we return error here, the node won't advance: this will keep the DB as consistent as possible. If we not, the node will advance, but the data might not be stored inside

File: storage/pruning/pruningStorer.go [hunk 12]
```diff
@@ -266,6 +267,60 @@ func (ps *PruningStorer) Put(key, data []byte) error {
...
+	}
+	defer closePersister()
+
+	return ps.doPutInPersister(key, data, persister)
+}
+
+func (ps *PruningStorer) createAndInitPersisterIfClosed(pd *persisterData) (storage.Persister, func(), error) {
+	isOpen := !pd.getIsClosed()
+	if isOpen {
+		noopClose := func() {}
```
Comment (iulianpascalau) [inline storage/pruning/pruningStorer.go:293]: noopClose for closing the persister. Smart

File: node/nodeTransactions_test.go [hunk 13]
```diff
@@ -310,13 +310,13 @@ func TestNode_PutHistoryFieldsInTransaction(t *testing.T) {
...
-func createNode(t *testing.T, withFullHistory bool) (*Node, *genericmocks.ChainStorerMock, *testscommon.PoolsHolderMock, *testscommon.HistoryRepositoryStub) {
-	chainStorer := genericmocks.NewChainStorerMock()
+func createNode(t *testing.T, epoch uint32, withDbLookupExt bool) (*Node, *genericmocks.ChainStorerMock, *testscommon.PoolsHolderMock, *testscommon.HistoryRepositoryStub) {
+	chainStorer := genericmocks.NewChainStorerMock(epoch)
 	dataPool := testscommon.NewPoolsHolderMock()
 	marshalizer := &mock.MarshalizerFake{}
 
 	historyRepo := &testscommon.HistoryRepositoryStub{
-		IsEnabledCalled: func() bool { return withFullHistory },
+		IsEnabledCalled: func() bool { return withDbLookupExt },
```
[the remark this answers is not in the document]
Comment (andreibancioiu) [reply node/nodeTransactions_test.go:319]: Fixed.

File: storage/factory/pruningStorerFactory.go [hunk 14]
```diff
@@ -481,6 +481,7 @@ func (psf *StorageServiceFactory) createPruningStorerArgs(storageConfig config.S
 		NumOfActivePersisters: numOfActivePersisters,
 		Notifier:              psf.epochStartNotifier,
 		MaxBatchSize:          storageConfig.DB.MaxBatchSize,
+		DbLookupExtensions:    psf.generalConfig.DbLookupExtensions.Enabled,
```
[the remark this answers is not in the document]
Comment (andreibancioiu) [reply storage/factory/pruningStorerFactory.go:484]: Fixed.

File: storage/pruning/pruningStorer.go (hunk 11 above)
Comment (andreibancioiu) [reply storage/pruning/pruningStorer.go:249]: Understood. Left as it were - to prevent breaking something unforeseen. However, let me know if I should change instead.

**Review by iulianpascalau — APPROVED** [review]

File: cmd/node/config/config.toml [hunk 15]
```diff
@@ -480,8 +480,8 @@
    Type = "json"
 
 [EpochStartConfig]
-    MinRoundsBetweenEpochs = 200
-    RoundsPerEpoch         = 200
+    MinRoundsBetweenEpochs = 100 # TODO UNDO AFTER TESTS
```
Comment (AdoAdoAdo) [inline cmd/node/config/config.toml:483]: undo ^^

File: storage/pruning/pruningStorer.go [hunk 16]
```diff
@@ -502,6 +517,7 @@ func (ps *PruningStorer) HasInEpoch(key []byte, epoch uint32) error {
 		}
 
 		persister, err := ps.persisterFactory.Create(pd.path)
+		// Question for review: save reference on persister data?
```
[the remark this answers is not in the document]
Comment (andreibancioiu) [reply storage/pruning/pruningStorer.go]: Ok.

File: cmd/node/config/config.toml (hunk 15 above)
Comment (andreibancioiu) [reply cmd/node/config/config.toml]: Fixed.

File: storage/pruning/pruningStorer.go (hunk 11 above)
Comment (bogdan-rosianu) [reply storage/pruning/pruningStorer.go]: this case didn't happen because the `epochForPutOperation` is set on `CommitBlock` so there are very small chances that the persister for the specified epoch doesn't exist. can remain like this for now

**Review by bogdan-rosianu — APPROVED** [review]

**Review by AdoAdoAdo — APPROVED** [review]

File: cmd/node/factory/structs.go [hunk 17]
```diff
@@ -628,20 +628,40 @@ func prepareGenesisBlock(args *processComponentsFactoryArgs, genesisBlocks map[u
...
+		if args.shardCoordinator.SelfId() != shard {
+			continue
+		}
+
+		genesisBlockHash, err := core.CalculateHash(args.coreData.InternalMarshalizer, args.coreData.Hasher, genesisBlockHeader)
+		if err != nil {
+			return err
+		}
+
+		log.Info("indexGenesisBlocks(): historyRepo.RecordBlock", "shard", shard, "hash", genesisBlockHash)
```
Comment (miiu96) [inline cmd/node/factory/structs.go:653]: is ok this log.Info ?  maybe debug ?

Comment (andreibancioiu) [reply cmd/node/factory/structs.go:653]: Indeed, I wanted it to be more obvious (since it was missing before).

**Review by miiu96 — APPROVED** [review]

**Review by LucianMincu — APPROVED** [review]
System tests passed.
